Initial 6 axis Support

tsmarks opened the pull request:
Added basic 6 axis features: DRO, Control
Added a 4th axis roughing strategy

As requested by creator, the DRO and Control panes can be removed/activated in the bCNC.ini file line:  control.page = DRO abcDRO State Control abcControl

Harvie commented:
Hello, thanks for your contribution. I still have to review the code, but i've tried to start your 6-Axis bCNC on my computer in both python2 and 3 and it seems to work. I also like the way you've separated additional axes into collapsible sections. Can you make it collapsed by default, so 3 axis users don't see it unless they click on it? So by default i looks like that:

![image](https://user-images.githubusercontent.com/276504/79305132-11435700-7ef3-11ea-96f2-11a8bc0c1561.png)

tsmarks commented:
Thanks.  I just added the extra axis to my machine and have been working on software to support it.  The first fix was reworking the grbl post processor for Fusion 360.  Now I figured get this GUI to support the basics.

I can look into that the default state.  I would assume, it would be even better to have the default bCNC.ini set to not display them at all.  Let the minority that want the 6 axis edit the ini to add them back in.

13: control.page = DRO abcDRO State Control abcControl
if it is:
13: control.page = DRO State Control
they don't show at all

Harvie commented:
> Let the minority that want the 6 axis edit the ini to add them back in.

Good idea...

Harvie commented:
BTW what is displayed on canvas when you load 6 axis g-code into bCNC? I guess that might be a problem...

tsmarks commented:
The visualization ignores the extra axis moves.  So for 4th axis operations you see a 2d profile of all the moves parallel to the rotary and up and down in Z.  It is useful for locating the job.  Fusion 360 struggles to visualize multi-axis in the simulation too.  I have thought about attacking the issue, but I fear it might be more resource intensive than people want bCNC to be.   The canvas would need to be a true 3d viewer to really make it work.

Harvie commented:
> I fear it might be more resource intensive than people want bCNC to be. The canvas would need to be a true 3d viewer to really make it work.

Oh yes. These are two greatest problems of bCNC... I am considering porting it to OpenGL or something... But i am not really sure how to do it and if it can help. Some people say that even with OpenGL it will be still slow, because drawed shapes are preprocessed in python anyway...

MARIOBASZ commented:
Very good tool to be able to handle 6 axes, soon I will add rotary axes to my machine, so I hope to use what you have done, thank you very much !!
The bcnc canvas is certainly a critical point (trochoidal use, the gcode have too many lines, it gets too slow, and / or graphics incomplete).
Regards.

tsmarks commented:
Mario,

I used a Arduino Mega with the Andy's CNC shield from [Eccentric Workshop](https://shop.eccentricworkshop.com/product-category/electronic-products/cnc-controllers/?v=7516fd43adaa) as a breakout and the GRBL build he provided too.  I had to rewrite Fusion 360's GRBL post processor to work with multi-axis, but now everything works well.  I have a few posts over on the [Fusion forum](https://forums.autodesk.com/t5/fusion-360-manufacture/grbl-multi-axis-post-processor/td-p/9413041) with the new post processor if you want the code.  I have successfully made a few 4th axis projects, but really needed this bCNC change to make it work smooth.  Hand jamming gcode for moving and zeroing the 4th was not friendly and having no DRO was a pain. (Not insulting bCNC, VERY happy with it, and thankful for the code being opensource.)

MARIOBASZ commented:
@tsmarks, I have on my pc still not using a version of grbl 6 axes for mega, fusion360 I have not used it yet ... all my objects have been designed with bcnc. Another advance for bcnc to be able to handle 6 axes.

I think that if you add in the dro to be able to choose, when it is commanded with the direction arrows, if the movement is G0 (like now) or G1, it can be useful, for example to go cutting almost manually. Hugs.

Harvie commented:
Can you please add some config checkbox for enabling 6axis mode? So i can merge this...

![image](https://user-images.githubusercontent.com/276504/79597121-06024e00-80e2-11ea-8e1e-dfd09d3a1741.png)

Harvie commented:
Maybe it would make sense to add few mentions of this to wiki. You should have write access:
https://github.com/vlachoudis/bCNC/wiki

Harvie commented:
Maybe it would make sense to add few mentions of this to wiki. You should have write access:
https://github.com/vlachoudis/bCNC/wiki

tsmarks commented:
I have added the check box in the config.  I also added a few blurbs in the Wiki and added apage for the Spiral plugin.  

Thanks for your interest in this.

Harvie commented:
Looks good to me, i will merge this. Do you have something you want to add before merge?

walterwissmann commented:
Many thanks to all to let more axis come to bCNC!

Walter

tsmarks commented:
Harvie, 
I can't think of anything.  If anything come up I will revisit.  Thanks!

Harvie commented:
Thank you for usefull feature! I hope the GUI will get bit more streamlined in the future. But i've never really worked on 6-axis machine, so i am no really aware about how that should look to be ergonomic...

TNO-CODE commented:
Thank you from the bottom of my heart 💓 .
Thanks to everyone who had input with this ,can't wait to try more Axis as I've already ordered the parts .

Harvie commented:
There is bug report claiming this had broken GRBLv1 compatibility... Can you please check it?
https://github.com/vlachoudis/bCNC/issues/1407

Harvie commented:
@tsmarks also can you please take a look at this? #1426

sttts commented on bCNC/controllers/_GenericController.py:
```diff
@@ -8,8 +8,8 @@
-STATUSPAT = re.compile(r"^<(\w*?),MPos:([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),WPos:([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),?(.*)>$")
-POSPAT	  = re.compile(r"^\[(...):([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*):?(\d*)\]$")
+STATUSPAT = re.compile(r"^<(\w*?),MPos:([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),WPos:([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),?(.*)>$")
+POSPAT	  = re.compile(r"^\[(...):([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*),([+\-]?\d*\.\d*):?(\d*)\]$")
```
the new groups in the pattern are non-optional, aren't they? With my GRBL I get `[PRB:-356.0,-5.28,-118.3,0.0:1]`, i.e. only 4 axes. What do I miss?

Actually, even the old pattern with only 3 coordinates is broken with my 4 axes GRBL.

Harvie replied:
Are you able to provide PR with fix please?

sttts replied:
https://github.com/vlachoudis/bCNC/pull/1441

sttts replied:
Not tested yet. 

Spent the afternoon hunting issues with z probing, giving wrong values. I hope this is was the root cause. Going to try now on real machine.
